library/core/src/cell.rs

```diff
@@ -1573,20 +1572,65 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 ///
 /// Note that while mutating or mutably aliasing the contents of an `&UnsafeCell<T>` is
 /// ok (provided you enforce the invariants some other way), it is still undefined behavior
-/// to have multiple `&mut UnsafeCell<T>` aliases.
+/// to have multiple `&mut UnsafeCell<T>` aliases. That is, `UnsafeCell` is a wrapper
+/// designed to have a special interaction (_w.r.t._ the compiler) with _shared_
+/// accesses (_i.e._, through an `&UnsafeCell<_>` reference): there is no magic whatsoever
+/// when dealing with _exclusive_ access (an `&mut UnsafeCell<_>`): neither the cell
+/// nor the wrapped value may be aliased for the duration of that `&mut` borrow.
+///
+///   - This is showcased by the [`.get_mut()`] accessor, which is a non-`unsafe`
+///     getter that yields a `&mut T`.
+///
+/// [`.get_mut()`]: `UnsafeCell::get_mut`
 ///
 /// # Examples
 ///
 /// ```
 /// use std::cell::UnsafeCell;
 ///
-/// # #[allow(dead_code)]
-/// struct NotThreadSafe<T> {
-///     value: UnsafeCell<T>,
-/// }
+/// let x: UnsafeCell<i32> = 42.into();
+/// // Get multiple / concurrent / shared references to the same `x`.
+/// let (p1, p2): (&UnsafeCell<i32>, &UnsafeCell<i32>) = (&x, &x);
+///
+/// unsafe {
+///     // SAFETY: within this scope there are no other references to `x`'s contents,
+///     // so ours is effectively unique.
+///     let p1_exclusive: &mut i32 = &mut *p1.get(); // -- borrow --+
+///     *p1_exclusive += 27; //                                     |
+/// } // <---------- cannot go beyond this point -------------------+
 ///
-/// unsafe impl<T> Sync for NotThreadSafe<T> {}
+/// unsafe {
+///     // SAFETY: within this scope nobody expects to have exclusive access to `x`'s contents,
+///     // so we can have multiple shared accesses concurrently.
+///     let p2_shared: &i32 = &*p2.get();
+///     assert_eq!(*p2_shared, 42 + 27);
+///     let p1_shared: &i32 = &*p1.get();
+///     assert_eq!(*p1_shared, *p2_shared);
+/// }
 /// ```
+///
+///   - Exclusive access to an `UnsafeCell<T>` ⇒ exclusive access to its `T`:
```

Comment:
These bullet points are rather strange, as there is a section title and a huge code block between them. They do not look loke an enumeration.

I think you can just make both of them regular sentences. However, if the second example has a title, then the first one should have a title, too.